File: AngularJS/src/org/angular2/lang/types/Angular2BaseType.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.types;

import com.intellij.lang.javascript.psi.JSType;
import com.intellij.lang.javascript.psi.JSTypeSubstitutionContext;
import com.intellij.lang.javascript.psi.JSTypeTextBuilder;
import com.intellij.lang.javascript.psi.JSTypeWithIncompleteSubstitution;
import com.intellij.lang.javascript.psi.types.JSCodeBasedType;
import com.intellij.lang.javascript.psi.types.JSSimpleTypeBaseImpl;
import com.intellij.lang.javascript.psi.types.JSTypeSource;
import com.intellij.lang.javascript.psi.types.JSTypeSourceFactory;
import com.intellij.psi.PsiElement;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public abstract class Angular2BaseType<T extends PsiElement> extends JSSimpleTypeBaseImpl
  implements JSCodeBasedType,                                        JSTypeWithIncompleteSubstitution {

  protected Angular2BaseType(@NotNull T source, Class<T> sourceClass) {
    this(JSTypeSourceFactory.createTypeSource(source, true), sourceClass);
  }

  protected Angular2BaseType(@NotNull JSTypeSource source, Class<T> sourceClass) {
    super(source);
    PsiElement sourceElement = source.getSourceElement();
    assert sourceElement != null;
    sourceClass.cast(sourceElement);
  }

  protected abstract @Nullable String getTypeOfText();

  protected abstract @Nullable JSType resolveType(@NotNull JSTypeSubstitutionContext context);

  @Override
  public @NotNull T getSourceElement() {
    //noinspection unchecked
    return (T)Objects.requireNonNull(super.getSourceElement());
  }

  @Override
  protected @Nullable JSType substituteImpl(@NotNull JSTypeSubstitutionContext context) {
    JSType type = resolveType(context);
    if (type != null) {
      context.add(type);
    }
    return type;
  }

  @Override
  public @NotNull JSType substituteCompletely() {
    return this.substitute().substitute();
  }

  @Override
  protected boolean isEquivalentToWithSameClass(@NotNull JSType type, @Nullable ProcessingContext context, boolean allowResolve) {
    return type.getClass() == this.getClass()
           && Objects.equals(type.getSourceElement(), getSourceElement());
  }

  @Override
  protected int resolvedHashCodeImpl() {
    return Objects.hash(getClass(), getSourceElement());
  }

  @Override
  protected void buildTypeTextImpl(@NotNull TypeTextFormat format, @NotNull JSTypeTextBuilder builder) {
    if (format == TypeTextFormat.SIMPLE) {
      builder.append("ngtypeof#" + getTypeOfText());
      return;
    }
    substitute().buildTypeText(format, builder);
  }
}

File: AngularJS/src/org/angular2/lang/types/Angular2BaseType.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.types

import com.intellij.lang.javascript.psi.JSType
import com.intellij.lang.javascript.psi.JSType.TypeTextFormat
import com.intellij.lang.javascript.psi.JSTypeSubstitutionContext
import com.intellij.lang.javascript.psi.JSTypeTextBuilder
import com.intellij.lang.javascript.psi.JSTypeWithIncompleteSubstitution
import com.intellij.lang.javascript.psi.types.JSCodeBasedType
import com.intellij.lang.javascript.psi.types.JSSimpleTypeBaseImpl
import com.intellij.lang.javascript.psi.types.JSTypeSource
import com.intellij.lang.javascript.psi.types.JSTypeSourceFactory
import com.intellij.psi.PsiElement
import com.intellij.util.ProcessingContext

abstract class Angular2BaseType<T : PsiElement?> protected constructor(source: JSTypeSource, sourceClass: Class<T>)
  : JSSimpleTypeBaseImpl(source), JSCodeBasedType, JSTypeWithIncompleteSubstitution {
  protected constructor(source: T, sourceClass: Class<T>) : this(JSTypeSourceFactory.createTypeSource(source, true), sourceClass)

  init {
    sourceClass.cast(source.sourceElement)
  }

  protected abstract val typeOfText: String?
  protected abstract fun resolveType(context: JSTypeSubstitutionContext): JSType?
  override fun getSourceElement(): T {
    @Suppress("UNCHECKED_CAST")
    return super<JSSimpleTypeBaseImpl>.getSourceElement()!! as T
  }

  override fun substituteImpl(context: JSTypeSubstitutionContext): JSType? {
    val type = resolveType(context)
    if (type != null) {
      context.add(type)
    }
    return type
  }

  override fun substituteCompletely(): JSType {
    return this.substitute().substitute()
  }

  override fun isEquivalentToWithSameClass(type: JSType, context: ProcessingContext?, allowResolve: Boolean): Boolean {
    return type.javaClass == this.javaClass && type.sourceElement == sourceElement
  }

  override fun hashCodeImpl(): Int {
    return sourceHashCode
  }

  override fun buildTypeTextImpl(format: TypeTextFormat, builder: JSTypeTextBuilder) {
    if (format == TypeTextFormat.SIMPLE) {
      builder.append("ngtypeof#$typeOfText")
      return
    }
    substitute().buildTypeText(format, builder)
  }
}